common-app-lib/src/cmr/common_app/api/log.clj

```diff
@@ -0,0 +1,133 @@
+(ns cmr.common-app.api.log
+  "Defines the log routes for applications. There are two main functions 1) To get the current
+   CMR logging configuration and 2) Change the logging configuration. To get the logging
+   configuration from a service such as search locally one would use the following command:
+   curl -H \"Echo-Token: XXXX\" http://localhost:3003/log where the port goes to the correct service.
+   From the internet the command is:
+   curl -H \"Echo-Token: XXXX\" http://cmr.earthdata.nasa.gov/search/log
+   The above commands return the current CMR logging configuration to the caller. It will format the
+   message to either a basic html if one of the accepted headers is text/html, or as formatted raw text
+   if text/html is not included in the accept HTML header.
+   To change the configuration one needs to provide an EDN map of the configuration you want to replace.
+   For example if the configuraiton looks like the following:
+   {:level :warn,
+    :ns-whitelist [],
+    :ns-blacklist [],
+    :middleware [#function[cmr.common.log/log-by-ns-pattern]],
+    :timestamp-opts
+    {:pattern \"yyyy-MM-dd HH:mm:ss.SSS\",
+     :locale :jvm-default,
+     :timezone :utc},
+    :output-fn #function[cmr.common.log/log-formatter],
+    :appenders
+    {:println
+     {:enabled? true,
+      :async? false,
+      :min-level nil,
+      :rate-limit nil,
+      :output-fn :inherit,
+      :fn
+      #function[taoensso.timbre.appenders.core/println-appender$fn--10435]}},
+    :ns-pattern-map
+    {:all :warn}}
+    use the following commmand:
+    curl -i -XPUT -H \"Echo-Token: XXXX\" -H \"Content-Type: application/edn\" <URL> -d
+        '{:level :debug :ns-pattern-map {\"cmr.common-app.api.log\" :debug :all :warn}}'
+    Where <URL> is either for example http://localhost:3003/log or
+     http://cmr.earthdata.nasa.gov/search/log
+    This command changes the overall logging level to debug - then sets the cmr.common-app.api.log
+    namespace to debug and all other names spaces are set to warn.
+    ******* Important! - The main logging level :level - must be set at the lowest setting that you
+    want logged at the namespace level. Otherwise Timbre will filter these logs out before the logs
+    make it to the middleware.  The middleware :all will set the logging level for all namespaces
```

Comment:
If someone leaves out :all from the map, could we default to adding :all and the current logging level?

Reply:
There is no way to take out :all from the API - there is no way to remove any key from the API, I have added a reset function that resets the logging back to what it was before a user changed it. All configuration changes are "merged" - added or changed only, never removed.